packages/cloud_firestore/cloud_firestore_web/lib/document_reference_web.dart

```diff
@@ -3,42 +3,40 @@ part of cloud_firestore_web;
 class DocumentReferenceWeb extends DocumentReference {
   final web.Firestore firestoreWeb;
   final web.DocumentReference delegate;
-  DocumentReferenceWeb(this.firestoreWeb,FirestorePlatform firestore, List<String> pathComponents)
-      : delegate = firestoreWeb.doc(pathComponents.join("/")),super(firestore, pathComponents);
+
+  DocumentReferenceWeb(this.firestoreWeb, FirestorePlatform firestore,
+      List<String> pathComponents)
+      : delegate = firestoreWeb.doc(pathComponents.join("/")),
+        super(firestore, pathComponents);
 
   @override
-  Future<void> setData(Map<String, dynamic> data, {bool merge = false}) {
-    return delegate.set(data, web.SetOptions(merge: merge));
-  }
+  Future<void> setData(Map<String, dynamic> data, {bool merge = false}) =>
+      delegate.set(FieldValueWeb._serverDelegates(data), web.SetOptions(merge: merge));
 
   @override
-  Future<void> updateData(Map<String, dynamic> data) {
-    return delegate.update(data: data);
-  }
+  Future<void> updateData(Map<String, dynamic> data) =>
+      delegate.update(data: FieldValueWeb._serverDelegates(data));
 
   @override
   Future<DocumentSnapshot> get({Source source = Source.serverAndCache}) async {
+    //TODO(amr): Honour source by passing it to the delegate
```

Comment:
Thanks for noticing!

Reply:
Actually, Firebase for web doesn't support passing a source. Now I am more inclined to ignore it here.